api/server/backend/build/containerbackend.go

```diff
@@ -0,0 +1,159 @@
+package build
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/builder"
+	containerpkg "github.com/docker/docker/container"
+	"github.com/docker/docker/pkg/stringid"
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+)
+
+// containerComponent is the daemon methods used by containerManager
+type containerComponent interface {
+	// ContainerAttachRaw attaches to container.
+	ContainerAttachRaw(cID string, stdin io.ReadCloser, stdout, stderr io.Writer, stream bool, attached chan struct{}) error
+	// ContainerCreate creates a new Docker container and returns potential warnings
+	ContainerCreate(config types.ContainerCreateConfig) (container.ContainerCreateCreatedBody, error)
+	// ContainerRm removes a container specified by `id`.
+	ContainerRm(name string, config *types.ContainerRmConfig) error
+	// ContainerKill stops the container execution abruptly.
+	ContainerKill(containerID string, sig uint64) error
+	// ContainerStart starts a new container
+	ContainerStart(containerID string, hostConfig *container.HostConfig, checkpoint string, checkpointDir string) error
+	// ContainerWait stops processing until the given container is stopped.
+	ContainerWait(ctx context.Context, name string, condition containerpkg.WaitCondition) (<-chan containerpkg.StateStatus, error)
+}
+
+type containerManager struct {
+	tmpContainers map[string]struct{}
+	backend       containerComponent
+}
+
+// NewContainerBackend creates a new container backend
+func NewContainerBackend(docker containerComponent) builder.ContainerBackend {
+	return &containerManager{
+		backend:       docker,
+		tmpContainers: make(map[string]struct{}),
```

Comment:
Yes, the entire purpose of this object is to manage the `tmpContainers`. If we remove them then we  are back to the old interface. 

I agree this is an incremental change which gets us closer to the final interface.

This was original part of the builder and you asked me to move it out! I can add a TODO on the struct explaining why it exists.